launch_ros/launch_ros/actions/load_composable_nodes.py

```diff
@@ -230,19 +233,48 @@ def execute(
         # Generate load requests before execute() exits to avoid race with context changing
         # due to scope change (e.g. if loading nodes from within a GroupAction).
         load_node_requests = []
+        autostart_actions = []
         for node_description in self.__composable_node_descriptions:
             request = get_composable_node_load_request(node_description, context)
             # The request can be None if the node description's condition evaluates to False
             if request is not None:
                 load_node_requests.append(request)
 
+            # If autostart is enabled, transition to the 'active' state.
+            if node_description.node_autostart:
+                complete_node_name = request.node_namespace + request.node_name
+                if not complete_node_name.startswith('/'):
+                    complete_node_name = '/' + complete_node_name
+                self.__logger.info(
+                    'Autostart enabled for requested lifecycle node {}'.format(complete_node_name))
+                node_description.init_lifecycle_event_manager(node_description, context)
+                autostart_actions.append(
+                    LifecycleTransition(
+                        lifecycle_node_names=[complete_node_name],
+                        transition_ids=[lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE]
+                    ))
+                autostart_actions.append(
+                    LifecycleTransition(
+                        lifecycle_node_names=[complete_node_name],
+                        transition_ids=[lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE]
+                    ),
+                )
+
```

Comment:
I won't block my review on this, but I wanted to point out that adding this lifecycle specific logic here isn't ideal. Instead, it would be great to have something like `on_execute(context) -> List[Action]` on `ComposableNode`, calling that for each `node_description` and accumulating the actions returned to be later returned at the end of this function as you're doing with `autostart_actions` but generically.

Then the `on_execute` in `ComposableNode` could be empty for now, and all of this highlighted logic could be moved into `LifecycleComposableNode`'s override of `on_execute()`. That way this class never has to understand anything about lifecycle nodes.

Reply:
I don't disagree it would nicely bin the lifecycle logic in discrete places, but the composable node / lifecycle composable nodes are both descriptions not actions. I think it would be against the design intent of the object to have execution-related methods like that. 

So, I think this can be left as-is?